sensor/common/registry/lazy_tls_factory.go

```diff
@@ -0,0 +1,95 @@
+package registry
+
+import (
+	pkgErrors "github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/registries"
+	"github.com/stackrox/rox/pkg/registries/docker"
+	dockerFactory "github.com/stackrox/rox/pkg/registries/docker"
+	rhelFactory "github.com/stackrox/rox/pkg/registries/rhel"
+	"github.com/stackrox/rox/pkg/registries/types"
+)
+
+var (
+	lazyEligibleCreators = []types.CreatorWrapper{
+		dockerFactory.CreatorWithoutRepoList,
+		rhelFactory.CreatorWithoutRepoList,
+	}
+)
+
+var _ registries.Factory = (*lazyFactory)(nil)
+
+type lazyFactory struct {
+	creators      map[string]types.Creator
+	tlsCheckCache *tlsCheckCacheImpl
+}
+
+func newLazyFactory(tlsCheckCache *tlsCheckCacheImpl) registries.Factory {
+	factory := &lazyFactory{
+		creators:      make(map[string]types.Creator, len(lazyEligibleCreators)),
+		tlsCheckCache: tlsCheckCache,
+	}
+
+	for _, creatorFunc := range lazyEligibleCreators {
+		typ, creator := creatorFunc()
+		factory.creators[typ] = creator
+	}
+
+	return factory
+}
+
+// CreateRegistry performs aggressive up front validation so that errors can
+// be surfaced, otherwise errors may be lost during lazy initialization.
+func (e *lazyFactory) CreateRegistry(source *storage.ImageIntegration, options ...types.CreatorOption) (types.ImageRegistry, error) {
```

Comment:
how come we don't use the options?

Reply:
That is a good catch, we should pass those down to the creator (interestingly style check didn't catch that) - will make that change.

I did initially consider adding a 'lazy init' option to the creator options, but abandoned that approach for multiple reasons - happy to walk through that decision if wanted, biggest reason is at the point 'creator' is called, everything is already initialized so there is nothing to lazy init / the option wouldn't help at that point.